### PR TITLE
Add server side previews to the application workflow

### LINKED ISSUE #3545: Display a server side preview of the application form before submitting
_**PLEASE DO NOT INCLUDE SENSITIVE INFORMATION**_

[Use this form](https://docs.google.com/forms/d/e/1FAIpQLSdcn68IgWuk5-YyXF8ZuVK5Dxj5az6eemmLlqDsghJwEmxY6A/viewform). if you need to include the following information:
- Account name or name of person
- Account email
- Any personally identifying information
- Title of submission (app URLs alone are OK)
- When in doubt, use the above form.

**Is your feature request related to a problem? Please describe.**

Yes. The richtext allows a copy-paste of the html content and preserves the markup. After submission when the content of the richtext is displayed, html is cleaned up and only a part of the support html is display. This leads to the removal of content that might be visible in the richtext editor. 

This also allows for an applicant to check and confirm the content and prevents accidentally submitting the application form. 

**Is your feature request related to an existing functionality? Please describe.**

Submisison and display on the html content in the application form.

**Describe the solution you'd like**

- Update the application submit button to "Preview and Submit", which after clicking doesn't immediately submits the form but allows the applicant to preview the form on a new page. On the preview page, the application form should be show as it would appear to the reviewers and staff. 

**Describe alternatives you've considered**

1. Add a preview button on the rich-text that would cleanup editor html server side and render cleanup content.
2. Fix the editor display to not display the html content that is not allowed. Essentially browser based bleach.
3. Use simplified markdown based editor with lightweigth formatting.

**Additional context**

User able to page in a image in the editor
image

The image is not render after it's submitted, if any goes to edit the application the image is displayed again.
image

**Priority**

- High (keeping you from completing day-to-day tasks)
- Medium (slowing you down)
- ✅ **Low priority (annoying, would be nice to not see)**

**Affected roles**

- Staff
- ✅ **Applicants**
- Reviewers
- Partners
- Developers
- General public

**Ideal deadline**

Date when you'd like to see this accomplished and a reason, if appropriate.

### PR DESCRIPTION
**Pull request by wes-otf**
## Description

Closes #3545.

This creates the option to require a user to preview their form before submitting it. The requirement is set on a per fund/lab basis, and is a ~required checkbox when the lab/fund is configured in Wagtail admin~ global Django configuration option. The preview first saves the application as a draft, then presents the user with a preview of their application. Upon viewing, the user can either edit or submit as seen in the flow diagram below.

*Sloppy Flow Diagram:*
Preview_Flow

*Screenshots*:
Screenshot 2024-01-12 at 18 42 15
Screenshot 2024-01-12 at 18 42 37

## Test Steps

**Testing preview on initial apply**
 - [ ] Navigate to https://test-apply.hypha.app/
 - [ ] Ensure you're logged in
 - [ ] Select `Apply` on any fund in the list
 - [ ] Populate some fields (ideally the ones that allow for text formatting)
 - [ ] Confirm `Preview and submit` appears at the bottom of the application
 - [ ] Select the `Preview and submit` button
 - [ ] Confirm a preview with the application's rendered fields now appears, with `Submit` and `Edit` buttons at the bottom.
 - [ ] Confirm selecting `Submit` successfully submits the submission (can be verified by using the navigation bar to go to `Submissions` > `All submissions`.

**Testing preview on edit**
 - [ ] Navigate to https://test-apply.hypha.app/
 - [ ] Ensure you're logged in
 - [ ] Select `Apply` on any fund in the list
 - [ ] Populate some fields (ideally the ones that allow for text formatting)
 - [ ] Select the `Save Draft` button at the bottom of the page
 - [ ] When redirected to the page stating`Your application is saved as a draft.`, select the `Continue editing` button.
 - [ ] Make changes to the application
 - [ ] Confirm the `Preview and submit` button is present at the bottom of the page.
 - [ ] Select the `Preview and submit` button
 - [ ] Confirm a preview with the updated application's rendered fields now appears, with `Submit` and `Edit` buttons at the bottom.
 - [ ] Confirm selecting `Submit` successfully submits the submission (can be verified by using the navigation bar to go to `Submissions` > `All submissions`.

### CODE REVIEW

**Comment by theskumar** [comment]
Do we really need a configuration option. I would suggest keeping it  as default for all the rounds/labs. 

One less configuration to care about.

**Comment by frjo** [comment]
This look like a nice implementation!

I would also prefer no option to make it more simple.

We can ask the other implementors and if some have really strong opinopn on it we might add a global env var to turn the previews off.

I can not see an organisation that wants previews for some fund/lab but not all.

**Comment by wes-otf** [comment]
Thanks @frjo @theskumar, this was great feedback! After chatting today in the maintainers meeting it seems like the path forward is global configurations for the previews rather than per-fund/lab as OTS brought up that some organizations might not want this feature. Will have those changes up soon!

File: hypha/settings/base.py [hunk 1]
```diff
@@ -562,6 +562,12 @@
 # Finance extension to finance2 for Project Invoicing
 INVOICE_EXTENDED_WORKFLOW = env.bool("INVOICE_EXTENDED_WORKFLOW", True)
 
+# Application Preview Settings
+
+# Require an applicant to view their rendered application before submitting
+REQUIRE_PREVIEW = env.bool("REQUIRE_PREVIEW", True)
```
Comment (frjo) [inline hypha/settings/base.py]: 1. Can we move this up to the "# Hypha custom settings" section. Good to have all the custom stuff in one place I think.
2. The setting name is a bit terse, could we do something like `SUBMISSION_PREVIEW_REQUIRED`? Something that tell what part of Hypha the setting refers to.

**Review by frjo — CHANGES_REQUESTED** [review]
This works really well in my testing. One small comment around the new env setting.

There is also an issue for sites that set:

```
FORCE_LOGIN_FOR_APPLICATION = False
```

The applicant fill in the submission form as anonymous (not logged in) and click "Preview and submit". They then have the option to click "Submit for review" or "Edit". Both will take them to the login screen without any explanations.

An account is created for them and the e-mail is sent out but we do not communicate this fact to them in Hypha.

If you set:

```
REQUIRE_PREVIEW = False
```

You still have the same issue if the applicant click the optional "Preview" button.

How should we handle this?

A simple way would be to make preview only work for logged in users.

**Comment by wes-otf** [comment]
@frjo thanks for testing! This is definitely an oversight from me, I'll go back to the drawing board for this one. 

> A simple way would be to make preview only work for logged in users.

I think this would also negate `FORCE_LOGIN_FOR_APPLICATION = False` if previewing was a piece of the submission process right? If you clicked "Preview and submit" and were forced to login to preview

**Comment by frjo** [comment]
@wes-otf I think the best solution is to only activate previewing for logged in users (regardless if `REQUIRE_PREVIEW` is set to true or false).

We then document that `FORCE_LOGIN_FOR_APPLICATION = False` will disable previewing of submissions for non logged in users.

What do you think?

**Comment by wes-otf** [comment]
@frjo Yeah I think that is the only answer that really makes sense. All the options I can think of would either be a major overhaul or a potential security vulnerability

**Comment by wes-otf** [comment]
This should be all set now!

**Review by frjo — APPROVED** [review]

**Comment by wes-otf** [comment]
@frjo I just created two labels for user testing vs dev testing. Let me know if you think that makes sense moving forward

**Pull request merged**